src/soundiness.rkt

```diff
@@ -45,22 +42,25 @@
   (match altn
     [(alt prog `(simplify ,loc ,input #f #f) `(,prev))
      (match-define (cons proof errors)
-       (cond
-         [(hash-has-key? simplify-cache input)
-          (hash-ref simplify-cache input)]
-         [else
-          (define proof
-            (get-proof input (location-get loc prog) (location-get loc (alt-program prev))))
-          (define vars (program-variables prog))
-          (cons proof (get-proof-errors proof pcontext ctx vars))
-          ]))
+       (hash-ref! simplify-cache input
+                  (λ ()
+                    (define proof (get-proof input
+                                             (location-get loc prog)
+                                             (location-get loc (alt-program prev))))
+                    ;; Proofs are actually on subexpressions,
+                    ;; we need to construct the proof for the full expression
+                    (define proof*
+                      (for/list ([step proof])
+                        (program-body (location-do loc prog (λ _ step)))))
+                    (define vars (program-variables prog))
+                    (cons proof* (get-proof-errors proof* pcontext ctx vars)))))
```

Comment:
Can you summarize the change here? There's a few things going on:

1. Adding things to the cache, surprised that didn't happen before.
2. Also some stuff with `location-do`, which I'm a little surprised by, what does that do?